### PR TITLE
refactor: use [Outside_build_dir.t]

### PR DESCRIPTION
**Pull request by rgrinberg**
Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

ps-id: 0b6059b1-b94a-4e8a-9e1c-68f9e86ea427

### CODE REVIEW

File: src/dune_engine/fs_memo.ml [hunk 1]
```diff
@@ -248,6 +249,7 @@ let path_kind path =
   | Error _ as error -> error
 
 let file_exists path =
+  let path = Path.as_outside_build_dir_exn path in
```
Comment (snowleopard) [inline src/dune_engine/fs_memo.ml:252]: Why are `file_exists` and a few other functions still taking `Path.t`? Why not switch the whole API to `Path.Outside_build_dir.t` consistently?

**Review by snowleopard — COMMENTED** [review]
I like where this is going, but why not convert the whole `fs_memo` API?

File: src/dune_engine/fs_memo.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_engine/fs_memo.ml:252]: It would be a much bigger change and perhaps the idea itself would be rejected. I can convert the rest of the api.

**Comment by rgrinberg** [comment]
@snowleopard okay, I changed the entire api.

File: src/dune_engine/fs_memo.ml [hunk 2]
```diff
@@ -13,7 +13,8 @@ module Watcher : sig
      directly if there is no parent. This is an optimisation that allows us to
      reduce the number of watched paths: typically, the number of directories is
      a lot smaller than the number of files. *)
-  val watch : try_to_watch_via_parent:bool -> Path.t -> unit Memo.t
+  val watch :
+    try_to_watch_via_parent:bool -> Path.Outside_build_dir.t -> unit Memo.t
 
   (* Invalidate a path after receiving an event from the file watcher. *)
   val invalidate : Path.t -> Memo.Invalidation.t
```
Comment (snowleopard) [inline src/dune_engine/fs_memo.ml:20]: Why not push the new type further and turn everything into `Path.Outside_build_dir` in the implementation too? Or is that going to be too invasive for some reason?

File: src/dune_engine/fs_memo.mli [hunk 3]
```diff
@@ -8,20 +8,23 @@ val init : dune_file_watcher:Dune_file_watcher.t option -> Memo.Invalidation.t
     build directory. *)
```
Comment (snowleopard) [inline src/dune_engine/fs_memo.mli:8]: This comment can now be dropped.

**Comment by snowleopard** [comment]
I think the new API in `fs_memo.mli` is a big improvement, thanks! As I mentioned in a comment, I wonder if we can push the new type a bit further inside the implementation, e.g. to have the Memo tables use `Path.Outside_build_dir.t` as well.

File: src/dune_engine/fs_memo.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_engine/fs_memo.ml]: I did this part of the API as well. There's still a bunch more places where we can tighten the types, but perhaps that should be subsequent work.

**Review by snowleopard — APPROVED** [review]
👍

**Pull request merged**